tiledb/sm/array/array.cc

```diff
@@ -523,6 +523,41 @@ Status Array::close() {
   return Status::Ok();
 }
 
+Status Array::delete_array(const URI& uri) {
+  // Check that query type is MODIFY_EXCLUSIVE
+  if (query_type_ != QueryType::MODIFY_EXCLUSIVE) {
+    return LOG_STATUS(Status_ArrayError(
+        "[Array::delete_array] Query type must be MODIFY_EXCLUSIVE"));
+  }
+
+  // Check that array is open
+  if (!is_open() && !controller().is_open(uri)) {
+    return LOG_STATUS(
+        Status_ArrayError("[Array::delete_array] Array is closed"));
+  }
+
+  // Check that array is not in the process of opening or closing
+  if (is_opening_or_closing_) {
+    return LOG_STATUS(Status_ArrayError(
+        "[Array::delete_array] "
+        "May not perform simultaneous open or close operations."));
+  }
+
+  // Delete array data
+  if (remote_) {
+    auto rest_client = storage_manager_->rest_client();
+    if (rest_client == nullptr) {
+      throw Status_ArrayError(
+          "[Array::delete_array] Remote array with no REST client.");
+    }
+    RETURN_NOT_OK(rest_client->delete_array_from_rest(uri));
+  } else {
+    RETURN_NOT_OK(storage_manager_->delete_array(uri.c_str()));
+  }
+
+  return Status::Ok();
```

Comment:
What is the state of the `Array` object at this point? There are no valid operations because even the schema should be gone. However, at this point the `Array` object does not (AFAICT) exclude a write query (allowed by MODIFY_EXCLUSIVE) -- the query might even "succeed", but the array would not be in a valid state.

Reply:
You bring up a great point; I've been trying to figure out the best solution here. 
It _may_ be sufficient to call `this->close()` before returning from `delete_array`: 

- enforce the array is closed for query & removed from the `Consistency` multimap
- reopen is disallowed by the `ArrayDirectory`;  schema has been deleted (array_directory.cc L195)
- users "re-invoking" `array.close()` after `array.delete_array` would be a no-op (array.cc L456)